### PR TITLE
Use stable harmonic angles, simultaneously turn on bonds and angles

### PR DESCRIPTION
**Pull request by mcwitt**
### Summary

Changes single topology transformations to
- interpolate bond and angle force constants simultaneously (instead of staggering)
- use `HarmonicAngleStable` (introduced in #935) instead of `HarmonicAngle`; this is necessary to ensure numerical stability when we turn on bonds and angles simultaneously
  - we use a small, positive $\epsilon$ for intermediate states only. End states have $\epsilon=0$, therefore reducing to the standard functional form

Also makes a small correction/simplification to the `HarmonicAngleStable` functional form: removing $\epsilon^2$ in the numerator, which was not necessary for the goal of numerical stability, and breaks a symmetry by biasing the numerator toward positive values.

### Motivation

Removing the staggering of bonds and angles
- is now safe with the introduction of a numerically stable harmonic angle potential
- decreases BAR errors of hif2a and tyk2 transformations (see [Benchmark results](#bench-results) below)
- simplifies optimal interpolation logic (e.g. makes unnecessary a proposed refinement that would e.g. disable staggering of angle terms that do not span a bond being formed)

#### hif2a

$\Delta \Delta G$ errors relative to reference (9e337214fe2b84fb0fd3c35df3c6fb6b5153a6b2):

|   mean |   std |   min |   25% |   50% |   75% |   max |
|-------:|------:|------:|------:|------:|------:|------:|
|  0.883 | 0.829 |  1.01 | 0.997 | 0.946 | 0.811 | 0.996 |

By transformation type:

| ring_transformation   |   mean |   std |   min |   25% |   50% |   75% |   max |
|:----------------------|-------:|------:|------:|------:|------:|------:|------:|
| closing               |  0.813 | 0.675 | 0.86  | 0.826 | 0.828 | 0.8   | 0.81  |
| none                  |  0.946 | 0.985 | 1.01  | 0.923 | 0.946 | 0.947 | 0.996 |
| opening               |  0.83  | 0.831 | 0.863 | 0.807 | 0.819 | 0.822 | 0.843 |

#### tyk2

|   mean |   std |   min |   25% |   50% |   75% |   max |
|-------:|------:|------:|------:|------:|------:|------:|
|  0.919 | 0.977 |     1 | 0.877 | 0.947 | 0.885 |  1.01 |

### CODE REVIEW

File: tests/test_velocity_verlet_integrator.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def test_reversibility():
     dt = 1.5e-3
 
     # Is not infinitely reversible, will fail after 3000 steps due to accumulation of coords/vels in floating point
-    for n_steps in [1, 10, 100, 500, 1000, 2000]:
+    for n_steps in [1, 10, 100, 500]:
```
Comment (maxentile) [inline tests/test_velocity_verlet_integrator.py:101]: max n_steps = 100 or 500 seems ample to me

(main motivation of this test was to assert suitability for use in HMC, which (1) is not yet incorporated in TM, (2) would not need to use infinitely long proposal trajectories, (3) could use an on-the-fly reversibility check if absolutely needed, as in 4.3 of https://arxiv.org/abs/2003.03950 (and references therein)... the previous large max n_steps may have been derived from tests on the reference implementation, which may be less impacted by precision loss)

File: timemachine/potentials/bonded_stable.py [hunk 2]
```diff
@@ -43,7 +43,7 @@ def harmonic_angle_stable(conf, params, angle_idxs, cos_angles=True):
     vij = ci - cj
     vkj = ck - cj
 
-    top = jnp.sum(vij * vkj, -1) + eps ** 2
+    top = jnp.sum(vij * vkj, -1)
```
Comment (mcwitt) [inline timemachine/potentials/bonded_stable.py:46]: The eps^2 term here was unnecessary for stability and introduced an asymmetry biasing the numerator in the positive direction.

File: timemachine/fe/single_topology.py [hunk 3]
```diff
@@ -1053,28 +1055,15 @@ def _setup_intermediate_chiral_bond_term(self, src_bond, dst_bond, lamb, interpo
...
-        """
+    def setup_intermediate_state(self, lamb):
+        r"""
         Setup intermediate states at some value of lambda.
 
         Parameters
         ----------
         lamb: float
-
-        lambda_angles, lambda_torsions: float
```
Comment (badisa) [inline timemachine/fe/single_topology.py:1064]: Some of the information in this comment seems like useful context, perhaps keeping the bits about how the interpolation works generally would be good.

File: timemachine/fe/single_topology.py [hunk 4]
```diff
@@ -681,7 +681,9 @@ def interpolate_harmonic_angle_params(src_params, dst_params, lamb, k_min, lambd
         lamb,
     )
 
-    return jnp.array([k, phase])
+    eps = jnp.where((lamb == 0.0) | (lamb == 1.0), 0.0, 1e-3)
```
Comment (badisa) [inline timemachine/fe/single_topology.py:684]: Should we have a constant for this epsilon value? Not sure how much giving it a different value matters or not.

**Comment by badisa** [comment]
Was there any impact on the BAR overlap going from the endpoints where `eps == 0` to intermediate simulations at `eps == 1e-3`?

**Comment by mcwitt** [comment]
@badisa there does not seem to be a significant difference in overlap between the end states and the neighboring intermediates when comparing the control and test versions:

imagea141-446c14741ff7.png)

I think this makes sense: under the assumption that end states are stable, we'd expect end state samples to be far away from singularities, in the region where `eps=1e-3` looks basically identical to `eps=0`.

File: timemachine/fe/single_topology.py (hunk 4 above)
Comment (mcwitt) [reply timemachine/fe/single_topology.py:684]: This was chosen somewhat arbitrarily to be large enough that no numerical instabilities were observed, and small enough to minimize the impact on end-state overlaps. Maybe we could add a comment to that effect? (Seemed premature to declare a constant, since I think this part of the code is relatively likely to change and it seemed useful to keep the definition local)

Comment (badisa) [reply timemachine/fe/single_topology.py:684]: A comment would work for me

File: timemachine/fe/single_topology.py (hunk 3 above)
Comment (mcwitt) [reply timemachine/fe/single_topology.py:1064]: Restored and expanded in https://github.com/proteneer/timemachine/pull/938/commits/8939e562b10f590e032e03be38371f2794fe7db4

File: timemachine/fe/single_topology.py (hunk 4 above)
Comment (mcwitt) [reply timemachine/fe/single_topology.py:688]: Added a comment in https://github.com/proteneer/timemachine/pull/938/commits/e52aa61342ef43b1d913fdfc3941e25713ee08ed

**Review by maxentile — APPROVED** [review]

**Review by badisa — APPROVED** [review]

**Review by jkausrelay — APPROVED** [review]
